_episodes/07-modules.md

```diff
@@ -22,6 +22,72 @@ This is what it looks like:
 import Trebuchet as Trebuchets
 using ForwardDiff: gradient
 
+mutable struct Trebuchet{T} <: AbstractVector{T}
+  counterweight::T
+  release_angle::T
+end
+Base.copy(trebuchet::Trebuchet) = Trebuchet(trebuchet.counterweight, trebuchet.release_angle)
+Base.size(trebuchet::Trebuchet) = tuple(2)
+Base.getindex(trebuchet::Trebuchet, i::Int) = getfield(trebuchet, i)
+function Base.setindex!(trebuchet::Trebuchet, v, i::Int)
+    if i === 1
+        trebuchet.counterweight = v
+    elseif i === 2
+        trebuchet.release_angle = v
+    else
+        error("Trebuchet only accepts indices 1 and 2, yours is $i")
+    end
+end
+
+struct Environment
+  wind::Float64
+  target_distance::Float64
+end
+
+function shoot_distance(windspeed, angle, weight)
+    Trebuchets.shoot(windspeed, angle, weight)[2]
+end
+function shoot_distance(args...)
+    Trebuchets.shoot(args...)[2]
+end
+function shoot_distance(trebuchet::Trebuchet, env::Environment)
+    shoot_distance(env.wind, trebuchet.release_angle, trebuchet.counterweight)
+end
+
+function aim(trebuchet::Trebuchet, environment::Environment; ε = 1e-1, η = 0.05)
+    better_trebuchet = copy(trebuchet)
+    hit = x -> (shoot_distance([environment.wind, x[2], x[1]]) - environment.target_distance)
```

Comment:
@mose do you know an easy way to preserve `Trebuchet`s under arithmetic? That is if `Trebuchet` is an `AbstractArray` `Trebuchet(1,2) + [3,4]` returns a `Trebuchet`?

I don't really want to cover promotion rules, but probably thats the only way out?

Reply:
What's `Trebuchets`?

Reply:
Maybe you don't need promotion rules, but first of all you're likely going to have to implement the [array interface](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array).  Then it's likely you can implement just `Base.:+(::Trebuchet, ::AbstractVector) = Trebuchet(...)`

Reply:
Thats three methods per operator for at least three operators. I guess that is also a bit too much.
Maybe the current version is not too bad after all

Reply:
I'll leave it for now